api/src/main/java/org/openmrs/module/appointmentscheduling/api/AppointmentService.java

```diff
@@ -980,4 +981,99 @@ Integer getHistoryCountByConditions(Date fromDate, Date endDate,
+	 */
+	@Authorized(AppointmentUtils.PRIV_VIEW_APPOINTMENT_RESOURCES)
+	List<AppointmentResource> getAllAppointmentResources(boolean includeVoided);
+
+	/**
+	 * Gets an appointment Resource by its appointment Resource id.
+	 *
+	 * @param appointmentResourceId the appointment Resource id.
+	 * @return the appointment Resource object found with the given id, else null.
+	 * @should get correct appointment Resource
+	 */
+	@Authorized(AppointmentUtils.PRIV_VIEW_APPOINTMENT_RESOURCES)
+	AppointmentResource getAppointmentResource(Integer appointmentResourceId);
+
+	/**
+	 * Gets an appointment Resource by its UUID.
+	 *
+	 * @param uuid the appointment Resource UUID.
+	 * @return the appointment Resource object found with the given uuid, else null.
+	 * @should get correct appointment Resource
+	 */
+	@Authorized()
+	AppointmentResource getAppointmentResourceByUuid(String uuid);
+
+	/**
+	 * Creates or updates the given appointment Resource in the database.
+	 *
+	 * @param appointmentResource the appointment Resource to create or update.
+	 * @return the created or updated appointment Resource.
+	 * @should save new appointment Resource
+	 * @should save a providerless appointment Resource
+	 * @should save edited appointment Resource
+	 * @should throw error when name is null
+	 * @should throw error when name is empty string
+	 */
+	@Authorized(AppointmentUtils.PRIV_MANAGE_APPOINTMENT_RESOURCES)
+	AppointmentResource saveAppointmentResource(AppointmentResource appointmentResource)
+			throws APIException;
+
+	/**
+	 * Voids a given appointment Resource.
+	 *
+	 * @param appointmentResource the appointment Resource to void.
+	 * @param reason              the reason why the appointment Resource is voided.
+	 * @return the appointment Resource that has been voided.
+	 * @should void given appointment Resource
+	 * @should void all associated time slots
+	 */
+	@Authorized(AppointmentUtils.PRIV_MANAGE_APPOINTMENT_RESOURCES)
+	AppointmentResource voidAppointmentResource(AppointmentResource appointmentResource,
+												String reason);
+
+	/**
+	 * Completely removes an appointment Resource from the database. This is not reversible.
+	 *
+	 * @param appointmentResource the appointment Resource to delete from the database.
+	 * @should delete given appointment Resource
+	 */
+	@Authorized(AppointmentUtils.PRIV_MANAGE_APPOINTMENT_RESOURCES)
+	void purgeAppointmentResource(AppointmentResource appointmentResource);
+
+	/**
+	 * Gets appointment Resource which have a given date, location, provider and list of appointment
+	 * types
+	 *
+	 * @return a list of appointment Resource objects.
+	 */
+	@Authorized(AppointmentUtils.PRIV_VIEW_APPOINTMENT_RESOURCES)
+	List<AppointmentResource> getAppointmentResourcesByConstraints(Location locations, Provider provider, List<AppointmentType> appointmentTypes);
+
+	@Authorized
+	TimeSlot getTimeslotForAppointment(Location location, Provider provider, AppointmentType type, Date appointmentDate);
```

Comment:
Instead of creating this new method, can't we sue the existing getTimeSlotByConstraints methods but with fromDate == toDate?

Reply:
This is not likely to work because the dates carry both the date and time and
1 . getTimeSlotByConstraints  compares using both the date and time,
fromDate==toDate will never return true because although the dates might be the same the start time and end time of an appointment are never  equal
2. Formatting the dates to exclude time in getTimeSlotByConstraints  will return more than one timeslot because a day can have multiple timeslot

Reply:
I see your point about 2) ... but maybe I'm not 100% understanding what you want to do...I'll add some further comments on that method.

Reply:
Could you just format the dates to exclude time and then if more than one time slot is returned just return the first one?

Reply:
from 2)  i don't just want to select the first timeslot, I want if the user schedules an appointment in the morning the morning timeslot gets selected ie 2019-06-10 7:00:00-2019-06-10 13:00:00 and if the appointment is in the afternoon the afternoon timeslot get returned by this getTimeslotForAppointment.

if I go with your suggested approach that means patients are all going to be scheduled to the morning timeslot and when it's full it's when the aftternoon timeslot is going to be selected, this is not what we want though

we want the flexibility to schedule patients in the morning and the afternoon timeslots independent of each other

Reply:
@mogoodrich  kindly check the my follow up comments and give me a go ahead